src/snowflake/cli/api/commands/snow_typer.py

```diff
@@ -111,3 +112,57 @@ def post_execute():
 
         log.debug("Executing command post execution callback")
         flush_telemetry()
+
+
+@dataclasses.dataclass
+class SnowTyperCommandData:
+    func: Callable
+    args: Tuple[Any, ...]
+    kwargs: Dict[str, Any]
+
+
+class SnowTyperCreator:
+    """
+    SnowTyper factory.
+    It is used to postpone operations needed for SnowTyper constructor,
+    which should not happen during library import (like config reading).
+
+    Usage:
+
+    ```
+    class MySnowTyperCreator(SnowTyperCreator):
+        def create_app(self) -> SnowTyper:
+            app = SnowTyper(...)
+            self.register_commands(app)
+            ...
+            return app
+
+    creator = MySnowTyperCreator()
+
+    @creator.command(...)
+    def actual_command(...):
+        ...
+    ```
+    """
+
+    def __init__(self):
+        self.commands_to_register: List[SnowTyperCommandData] = []
+
+    def create_app(self) -> SnowTyper:
+        """
+        Returns SnowTyper instance.
+        """
+        raise NotImplementedError()
+
+    def command(self, *args, **kwargs):
+        def decorator(command):
+            self.commands_to_register.append(
+                SnowTyperCommandData(command, args=args, kwargs=kwargs)
+            )
+            return command
+
+        return decorator
+
+    def register_commands(self, app: SnowTyper) -> None:
+        for command in self.commands_to_register:
+            app.command(*command.args, **command.kwargs)(command.func)
```

Comment:
Why we cannot add this logic to `SnowTyper`? What's reasoning behind adding this additional layer?

Reply:
We want to postpone calling `__init__` until error handling would be set up. The problem behind the issue was calling feature flags (which read config.toml) to determine constructor arguments (`hidden=FeatureFlag.is_disabled`)

Reply:
changed as discussed